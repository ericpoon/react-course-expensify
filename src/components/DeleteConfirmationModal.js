import React from 'react';
import Modal from 'react-modal';
import moment from 'moment/moment';
import numeral from 'numeral';

export default class DeleteConfirmationModal extends React.Component {
    render() {
        const {expense, isOpen, onRemoveConfirm, onRemoveCancel} = this.props;
        return (
            <Modal
                isOpen={isOpen}
                onRequestClose={onRemoveCancel}
                className={'modal'}
                portalClassName={'modal__wrapper'}
                closeTimeoutMS={200}
                appElement={document.getElementById('app')}
            >
                <div className={'modal__body'}>
                    <h3>Removing an expense</h3>
                    <p>Are you sure you want to remove the expense for <span>{expense.description}</span>?
                    </p>
                    <p>It's {numeral(expense.amount).format('$0,0.00')} and spent
                        on {moment(expense.createdAt).format('MMM D, YYYY')}.</p>
                </div>
                <div className={'modal__actions'}>
                    <button onClick={onRemoveConfirm} className={'button button--danger'}>Confirm</button>
                    <button onClick={onRemoveCancel} className={'button button--secondary'}>Cancel</button>
                </div>
            </Modal>
        );
    }
}