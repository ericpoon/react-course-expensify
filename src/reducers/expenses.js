const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE': {
      return [...state, action.expense];
    }
    case 'REMOVE_EXPENSE': {
      return state.filter(s => s.id !== action.id);
    }
    case 'EDIT_EXPENSE': {
      const { description, note, amount } = action.updates;
      const updates = { description, note, amount };
      return state.map((s) => {
        if (s.id === action.id) return { ...s, ...updates };
        return s;
      });
    }
    case 'SET_EXPENSES': {
      return action.expenses;
    }
    default: {
      return state;
    }
  }
};
