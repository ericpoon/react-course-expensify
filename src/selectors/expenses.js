import moment from 'moment';

export const selectFilteredExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses
        .filter(expense => {
            const {createdAt, description, amount} = expense;
            const createdAtMoment = moment(createdAt);
            const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
            const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
            const textMatch = typeof text !== 'string' || description.toLowerCase().includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a, b) => {
            switch (sortBy) {
            case 'amount':
                return a.amount < b.amount; // descending
            case 'date':
                return a.createdAt < b.createdAt; // descending
            default:
                return false;
            }
        });
};

export const selectExpensesTotal = (expenses) => {
    return expenses
        .map(ex => ex.amount)
        .reduce((sum, value) => sum + value, 0);
};