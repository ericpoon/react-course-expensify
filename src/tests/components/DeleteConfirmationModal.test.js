import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses'
import DeleteConfirmationModal from '../../components/DeleteConfirmationModal';

let wrapper, onRemoveConfirm, onRemoveCancel;

beforeEach(() => {
    const expense = expenses[0];
    onRemoveConfirm = jest.fn();
    onRemoveCancel = jest.fn();
    wrapper = shallow(
        <DeleteConfirmationModal
            isOpen={true}
            expense={expense}
            onRemoveConfirm={onRemoveConfirm}
            onRemoveCancel={onRemoveCancel}
        />,
    );
});

it('should render DeleteConfirmationModal correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

it('should do onRemoveConfirm after clicking confirm', () => {
    wrapper.find('button').at(0).simulate('click');
    expect(onRemoveConfirm).toHaveBeenCalled();
});

it('should do onRemoveCancel after clicking cancel', () => {
    wrapper.find('button').at(1).simulate('click');
    expect(onRemoveCancel).toHaveBeenCalled();
});