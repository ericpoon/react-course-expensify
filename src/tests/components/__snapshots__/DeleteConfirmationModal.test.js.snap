// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DeleteConfirmationModal correctly 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  isOpen={true}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="modal__wrapper"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    className="modal__body"
  >
    <h3>
      Removing an expense
    </h3>
    <p>
      Are you sure you want to remove the expense for 
      <span>
        Gum
      </span>
      ?
    </p>
    <p>
      It's 
      $195.00
       and spent on 
      Jan 1, 1970
      .
    </p>
  </div>
  <div
    className="modal__actions"
  >
    <button
      className="button button--danger"
      onClick={[MockFunction]}
    >
      Confirm
    </button>
    <button
      className="button button--secondary"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </div>
</Modal>
`;
